core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -452,6 +466,11 @@ private[log] class Cleaner(val id: Int,
                 }
                 else messageAndOffset
               }
+              // We need the max timestamp and last offset for time index
+              if (retainedMessages.last.message.timestamp > maxTimestamp) {
+                maxTimestamp = retainedMessages.last.message.timestamp
+                offsetOfMaxTimestamp = retainedMessages.last.offset
+              }
```

Comment:
Since we changed the logic, so for compressed messages, we need the offset of the wrapper message for the timeindex. Also, could we save the last MessageAndOffset to avoid calling retainedMessages.last (since it requires iteration)?